src/common/actions/repository.js

```diff
@@ -16,104 +14,80 @@ export function toggleRepositorySelection(id) {
+      }));
+
       if (owner) {
         dispatch(push(`/user/${owner}`));
       }
-    });
+    } catch(error) {
+      dispatch(addRepoError(error.action.id, error));
+    }
   };
 }
 
-// add a repository to launchpad's build queue
-export function addRepo(repository) {
-  const { id, url, owner, name } = repository;
+function addRepo(repository) {
+  const { id, url } = repository;
 
-  return async (dispatch) => {
-    dispatch({
-      type: REPO_ADD,
-      payload: {
-        id
-      }
-    });
-
-    try {
-      await createWebhook(owner, name);
-
-      // XXX
-      // actual LP API call that we use on server side returns a snap representation
-      // that we could use to immediatelly update client side state with newly
-      // added snap
-      const response = await fetch(`${BASE_URL}/api/launchpad/snaps`, {
+  return {
+    id,
+    [CALL_API]: {
+      path: '/api/launchpad/snaps',
+      types: [REPO_ADD_REQUEST, REPO_ADD_SUCCESS],
+      options: {
         method: 'POST',
         headers: { 'Content-Type': 'application/json' },
         body: JSON.stringify({ repository_url: url }),
         credentials: 'same-origin'
-      });
-
-      await checkStatus(response);
-      dispatch(addRepoSuccess(id));
-    } catch (error) {
-      dispatch(addRepoError(id, error));
-      return Promise.reject(error);
+      }
     }
   };
 }
 
-export function resetRepository(id) {
+function addRepoError(id, error) {
   return {
-    type: REPO_RESET,
+    type: REPO_ADD_FAILURE,
     payload: {
-      id
-    }
+      id,
+      error: new APICompatibleError({
+        code: error.json.body.code,
+        message: 'Repo cannot be added at this time',
+        detail: `Failed to add repo: ${error}`
+      })
+    },
+    error: true
   };
 }
 
-export async function createWebhook(owner, name) {
-  const response = await fetch(`${BASE_URL}/api/github/webhook`, {
-    method: 'POST',
-    headers: { 'Content-Type': 'application/json' },
-    body: JSON.stringify({ owner, name }),
-    credentials: 'same-origin'
-  });
-  if (response.status >= 200 && response.status < 300) {
-    return response;
-  } else {
-    const json = await response.json();
-    if (json.payload && json.payload.code === 'github-already-created') {
-      return response;
-    }
-    throw getError(response, json);
-  }
-}
-
-function addRepoSuccess(id) {
+export function resetRepository(id) {
   return {
-    type: REPO_SUCCESS,
+    type: REPO_RESET,
     payload: {
       id
     }
   };
 }
 
-function addRepoError(id, error) {
+function createWebhook({ owner, name }) {
   return {
-    type: REPO_FAILURE,
-    payload: {
-      id,
-      error
-    },
-    error: true
+    [CALL_API]: {
+      path: '/api/github/webhook',
+      options: {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({ owner, name }),
+        credentials: 'same-origin'
+      }
+    }
```

Comment:
No action types supplied. We don't currently handle problems installing single webhooks, but when a call results in an error it returns a rejected promise that is caught in [addRepos](https://github.com/canonical-ols/build.snapcraft.io/pull/635/files#diff-3d0e5cd44d7df7fc7a2fecb29c4a6deaR33)